code/controllers/admin.dm

```diff
@@ -59,11 +59,14 @@ ADMIN_VERB(debug_controller, R_DEBUG, "Debug Controller", "Debug the various per
 	var/list/controllers = list()
 	var/list/controller_choices = list()
 
-	for (var/datum/controller/controller in world)
+	for (var/var_key in global.vars)
```

Comment:
for clarity the proper way of doing what we USED to do (scan all datums in the game) would be 
```
for(var/datum/controller/controller)
```
this differs from what you're currently doing because it scans everywhere, not just globals (unsure if this is desired behavior or not)

Reply:
there has to be a cheaper way to do that

Reply:
@LemonInTheDark FWIW, all non-subsystem controllers (that is, instances of `subtypesof(/datum/controller) - subtypesof(/datum/controller/subsystem)`) are contained in global variables (actual global variables, not the `GLOB` type of global variables we use). Because of that, it's much less iteration to just iterate over all the variables under `global`.

That said, I can still change it to your code, if you want me to.

@Iamgoofball Unfortunately not - unless you know of a global variable that contains all non-subsystem controllers?

Reply:
> there has to be a cheaper way to find all controllers

it doesn't *really* matter since this verb is only invoked once in a blue moon

Reply:
Then we should definitely defer the work to initialization and then cache the results from there.

Reply:
> Then we should definitely defer the work to initialization and then cache the results from there.

A cache is way overkill here.

This is a debug verb - it's meant to be used on local debugging sessions, where it doesn't matter if you lag for a bit.

It might be used on live servers, sure - but not nearly as often as on local sessions. And if we *need* to use it on a live server, something has gone horribly wrong and we should probably be rebooting the server anyways.

Reply:
ftr:
this is so userspace you have no idea, you are correct a cache would be a waste (also risks issues if we have to rebuild some things)
you are correct searching global is way faster, we have very few things in there outside statics
I asked to ensure behavior was the same, which it seems to be so this'll do. it's a little more fineky but this verb IS THAT already so eh

Reply:
this is technically possible if the mc's failsafe system decides it's fucked and recreates it, and qdel's the old SS (this would lead to two of the same thing floating around in memory, at least until the old MC has all its refs cleared). I'd say that's a bit out of scope tho, and anyone who needs that level of detail can sdql for it